crates/sargon/src/signing/authentication/auth_intent.rs

```diff
@@ -0,0 +1,195 @@
+use crate::prelude::*;
+use std::hash::Hasher;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AuthIntent {
+    /// The challenge nonce that with some `metadata` values are generating the `RolaChallenge`
+    /// needed to be signed
+    pub challenge_nonce: Exactly32Bytes,
+
+    /// The `NetworkID` on which the request was made
+    pub network_id: NetworkID,
+
+    /// The origin `Url` of the dApp from which the request was made
+    pub origin: Url,
+
+    /// The dApp's definition address
+    pub dapp_definition_address: DappDefinitionAddress,
+
+    /// The entities needed to be signed.
+    pub entities_to_sign: Vec<AddressOfAccountOrPersona>,
+}
+
+impl AuthIntent {
+    /// Creates an `AuthIntent` from the request received from a dApp.
+    ///
+    /// Fails if the `OriginUrl` in metadata is not a valid url or the network id supplied does not
+    /// match the network id of the given dApp definition address or the entities to sign.
+    pub fn new_from_request(
+        challenge_nonce: DappToWalletInteractionAuthChallengeNonce,
+        metadata: DappToWalletInteractionMetadata,
+        entities_to_sign: Vec<AddressOfAccountOrPersona>,
+    ) -> Result<Self> {
+        let origin = TryInto::<Url>::try_into(metadata.origin.clone())?;
+
+        if metadata.network_id != metadata.dapp_definition_address.network_id()
+        {
+            return Err(CommonError::NetworkDiscrepancy {
+                expected: metadata.network_id,
+                actual: metadata.dapp_definition_address.network_id(),
+            });
+        }
+
+        for entity in &entities_to_sign {
+            if entity.network_id() != metadata.network_id {
+                return Err(CommonError::NetworkDiscrepancy {
+                    expected: metadata.network_id,
+                    actual: entity.network_id(),
+                });
+            }
+        }
```

Comment:
Added some validation when generating an `AuthIntent`. I think this is what we need.